### PR TITLE
Multi audio

### PR DESCRIPTION
**Pull request by AlinaVoronkovaEpam**

### CODE REVIEW

**Comment by AlinaVoronkovaEpam** [comment]
This PR is not fully completed.
Please, take a look, leave comments. Thanks.

File: sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java [hunk 1]
```diff
@@ -296,6 +297,17 @@ private void injectLocalizedResources(JSONObject configJson, Context context) {
     return languageFiles;
   }
 
+  private void setDefaultAudioLanguage(JSONObject audioLanguage) {
```
Comment (pilievOoyala) [inline sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java]: Please document all your methods. Even if they are private.  It's in our code style guide and will avoid making more code debt.

Comment (AlinaVoronkovaEpam) [reply sdk/android/skin/src/main/java/com/ooyala/android/skin/OoyalaSkinLayoutController.java]: Sure. Thanks. I'll do it on Monday

File: sdk/react/constants.js [hunk 2]
```diff
@@ -27,7 +27,7 @@ var BUTTON_NAMES = { // must match Objective-C code expectations & skin.json.
   // more option buttons
   DISCOVERY: 'discovery',
   QUALITY: 'quality',
-  CLOSED_CAPTIONS: 'closedCaption',
+  AUDIO_AND_CC: 'audioAndCC',
```
Comment (brol1dev) [inline sdk/react/constants.js:30]: Is this read correctly by the accessibility screen readers? I'm fearing it won't be understandable.

**Review by brol1dev — COMMENTED** [review]
I see some elements are enabled for accessibility. Did you test it? Is everything spelled correctly by the screen reader?

File: sdk/iOS/OoyalaSkinSDK/OOSkinPlayerObserver.m [hunk 3]
```diff
@@ -183,26 +186,26 @@ - (void) bridgeCurrentItemChangedNotification:(NSNotification *)notification {
...
-    @"promoUrl":promoUrl,
-    @"hostedAtUrl": hostedAtUrl,
-    @"duration":durationNumber,
-    @"live":live,
-    @"languages":closedCaptionsLanguages,
-    @"width":frameWidth,
-    @"height":frameHeight,
-    @"volume": volume};
-  [self.viewController sendBridgeEventWithName:notification.name body:eventBody];
+  NSMutableDictionary *eventBody = [[NSMutableDictionary alloc] initWithDictionary:
```
Comment (brol1dev) [inline sdk/iOS/OoyalaSkinSDK/OOSkinPlayerObserver.m:196]: Can we change it back to a `NSDictionary`?

File: sdk/react/bottomOverlay.js [hunk 4]
```diff
@@ -265,22 +267,27 @@ If the playhead position has changed, reset the cachedPlayhead to -1 so that it
...
+        volume={this.props.volume}
+        live={this.props.live}
+        width={this.props.width - 2 * leftMargin}
+        height={this.props.height}
+        fullscreen = {this.props.fullscreen}
+        onPress={this.props.onPress}
+        handleControlsTouch={this.props.handleControlsTouch}
+        showWatermark={this.props.showWatermark}
+        config={this.props.config}
+        сlosedCaptionsEnabled={this.props.сlosedCaptionsEnabled}
```
Comment (brol1dev) [inline sdk/react/bottomOverlay.js:286]: why did you add `closedCaptionsEnabled`?

Comment (brol1dev) [reply sdk/react/bottomOverlay.js:286]: In the UI we have two sections, how for audio tracks and one for CC.

File: sdk/react/controlBar.js [hunk 5]
```diff
@@ -211,6 +207,12 @@ var ControlBar = React.createClass({
         style: [styles.icon, {"fontSize": iconFontSize}, this.props.config.controlBar.iconStyle.active],
         icon: this.props.config.icons.stereoscopic
       },
+      audioAndCC: {
+        onPress: this.onAudioAndCCPress,
+        iconTouchableStyle: styles.iconTouchable,
+        style: [styles.icon, {"fontSize": iconFontSize}, this.props.config.controlBar.iconStyle.active],
+        icon: this.props.config.icons.audioAndCC,
```
Comment (brol1dev) [inline sdk/react/controlBar.js:214]: Note for everyone: We need to merge https://github.com/ooyala/skin-config/pull/195 for this line to work.

File: sdk/react/index.android.js [hunk 6]
```diff
@@ -57,6 +57,8 @@ var OoyalaSkin = React.createClass({
       // things which default to null and thus don't have to be stated:
       // selectedLanguage: null,
       // availableClosedCaptionsLanguages: null,
+      // selectedAudioTrack,
```
Comment (brol1dev) [inline sdk/react/index.android.js:60]: can you add `multiAudioEnabled: false` to the `getInitialState` function? So that we can see that at the start it is clear that multiAudio is not enabled.

File: sdk/react/index.ios.js [hunk 7]
```diff
@@ -50,6 +50,8 @@ var OoyalaSkin = React.createClass({
       // things which default to null and thus don't have to be stated:
       // selectedLanguage: null,
       // availableClosedCaptionsLanguages: null,
+      // selectedAudioTrack,
```
Comment (brol1dev) [inline sdk/react/index.ios.js:53]: can you add multiAudioEnabled: false to the getInitialState function? So that we can see that at the start it is clear that multiAudio is not enabled.

File: sdk/react/ooyalaSkinCore.js [hunk 8]
```diff
@@ -67,14 +67,20 @@ OoyalaSkinCore.prototype.handleLanguageSelection = function(e) {
...
+
 // event handlers.
 OoyalaSkinCore.prototype.handleMoreOptionsButtonPress = function(buttonName) {
   switch (buttonName) {
     case BUTTON_NAMES.DISCOVERY:
       this.pushToOverlayStackAndMaybePause(OVERLAY_TYPES.DISCOVERY_SCREEN);
       break;
-    case BUTTON_NAMES.CLOSED_CAPTIONS:
-      this.pushToOverlayStackAndMaybePause(OVERLAY_TYPES.CLOSEDCAPTIONS_SCREEN);
+    case BUTTON_NAMES.AUDIO_AND_CC:
```
Comment (brol1dev) [inline sdk/react/ooyalaSkinCore.js:82]: this is appearing twice, remove one of them.

File: sdk/react/panels/AudioAndCCSelectionPanel.js [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+  onAudioTrackSelected: function(name) {
+    if (this.props.selectedAudioTrackTitle !== name) {
+      this.props.onSelectAudioTrack(name);
+    }
+  },
+
+  onClosedCaptionsLanguageSelected: function(name) {
+    var offButtonLocalizedTitle = Utils.localizedString(this.props.config.locale, "Off", this.props.config.localizableStrings);
+
+    if (name == offButtonLocalizedTitle) {
```
Comment (brol1dev) [inline sdk/react/panels/AudioAndCCSelectionPanel.js:67]: use the `===` equality

Comment (brol1dev) [reply sdk/react/panels/AudioAndCCSelectionPanel.js:67]: @IvanSakharovskii, can you add a comment about why it is `==` necessary? So that someone who is new to this code understands why you used it this way.

File: sdk/react/panels/AudioAndCCSelectionPanel.js [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+  onDismissPress: function() {
+    this.props.onDismiss();
+  },
+
+  renderHeaderView: function(hasMultiAudioTracks, hasClosedCaptions) {
+    var leftTitle;
+    var rightTitle;
+
+    var localizedAudioTitle = Utils.localizedString(this.props.config.locale, "Audio", this.props.config.localizableStrings);
+    var localizedSubtitlesTitle = Utils.localizedString(this.props.config.locale, "Subtitles", this.props.config.localizableStrings);
```
Comment (akshaydch11) [inline sdk/react/panels/AudioAndCCSelectionPanel.js:83]: We are using Closed Caption as naming convention. Can we continue doing that instead "Subtitles" ?

File: sdk/react/panels/AudioAndCCSelectionPanel.js [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+
+    if (hasMultiAudioTracks && hasClosedCaptions) {
+       leftTitle = localizedAudioTitle;
+       rightTitle = localizedSubtitlesTitle;
+    } else if (hasMultiAudioTracks) {
+      leftTitle = localizedAudioTitle;
+      rightTitle = "";
+    } else {
+      leftTitle = localizedSubtitlesTitle;
```
Comment (akshaydch11) [inline sdk/react/panels/AudioAndCCSelectionPanel.js:92]: same as above

File: sdk/react/panels/ItemSelectionScrollView.js [hunk 12]
```diff
@@ -0,0 +1,78 @@
...
+    items: React.PropTypes.array,
+    selectedItem: React.PropTypes.string,
+    onSelect: React.PropTypes.func,
+    width: React.PropTypes.number,
+    height: React.PropTypes.number,
+    config: React.PropTypes.object
+  },
+
+  isSelected: function(name) {
+    return name && name !== '' && name == this.props.selectedItem;
```
Comment (brol1dev) [inline sdk/react/panels/ItemSelectionScrollView.js:29]: can you comment why you are using `==` here instead of `===`, if it is necessary, add a comment in line please.

**Review by brol1dev — CHANGES_REQUESTED** [review]
For this review, I checked react and iOS. I still need to see the changes for Android/Java.

Please update where necessary.

**Comment by pilievOoyala** [comment]
@AlinaVoronkovaEpam  @IvanSakharovskii Please address the comments in another PR.

**Review by pilievOoyala — APPROVED** [review]
